app/services/experiments/copy_experiment_as_template_service.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+module Experiments
+  class CopyExperimentAsTemplateService
+    extend Service
+
+    attr_reader :errors, :c_exp
+    alias cloned_experiment c_exp
+
+    def initialize(experiment_id:, project_id:, user_id:)
+      @exp = Experiment.find experiment_id
+      @project = Project.find project_id
+      @user = User.find user_id
+      @original_project = @exp&.project
+      @c_exp = nil
+      @errors = {}
+    end
+
+    def call
+      return self unless valid?
+
+      ActiveRecord::Base.transaction do
+        @c_exp = Experiment.new(
+          name: find_uniq_name,
+          description: @exp.description,
+          created_by: @user,
+          last_modified_by: @user,
+          project: @project
+        )
+
+        # Copy all signle taskas
+        @c_exp.my_modules << @exp.my_modules.without_group.map do |m|
+          m.deep_clone_to_experiment(@user, @c_exp)
+        end
+
+        # Copy all grouped tasks
+        @exp.my_module_groups.each do |g|
+          @c_exp.my_module_groups << g.deep_clone_to_experiment(@user, @c_exp)
+        end
+
+        raise ActiveRecord::Rollback unless @c_exp.save
+      end
+      @errors.merge!(@c_exp.errors.to_hash) unless @c_exp.valid?
+
+      @c_exp = nil unless succeed?
+      @c_exp.delay.generate_workflow_img if succeed?
+      track_activity if succeed?
+
+      self
+    end
+
+    def succeed?
```

Comment:
This could be in parent class.

Reply:
Create a parent class `MoveCloneService` which extends `Service.rb` then this one and the `MoveService` extend the `MoveCloneService`.  `succeed?` is then placed in the `MoveCloneService`, along with every common logic.

Your approach would probably be suitable if we had `@errors` and `succeed?` in every service.

I hope I'm not missing something here, I feel a bit under the weather.

Reply:
As you said. But yes, `errors` and `succeed?` should be in every service, I guess?